main.cc

```diff
@@ -84,30 +84,47 @@ void register_tracing_keyspace_backend(backend_registry& br);
 
 }
 
+// Must live in a seastar::thread
 class stop_signal {
     bool _caught = false;
     condition_variable _cond;
+    sharded<abort_source, plain_sharded_traits<abort_source>> _abort_sources;
+    //future<> _broadcasts_to_abort_sources_done = broadcast_loop();
+    future<> _broadcasts_to_abort_sources_done = make_ready_future<>();
 private:
     void signaled() {
+        if (_caught) {
+            return;
+        }
         _caught = true;
         _cond.broadcast();
+        _broadcasts_to_abort_sources_done = _broadcasts_to_abort_sources_done.then([this] {
+            startlog.info("broadcasting abort");
+            return _abort_sources.invoke_on_all(&abort_source::request_abort);
+        });
     }
 public:
     stop_signal() {
+        _abort_sources.start().get();
         engine().handle_signal(SIGINT, [this] { signaled(); });
         engine().handle_signal(SIGTERM, [this] { signaled(); });
     }
     ~stop_signal() {
+        _cond.broadcast();
```

Comment:
Not needed. It is a remnant of a different mechanism to stop the loop that broadcasted aborts to all shards (in the end, I found that I had if (!_caught) { return } instead of if (_caught) { return }, and that the original broadcast mechanism was fine.

I'll remove this bit.